### PR TITLE
Embed CLI into editor packages and update tests

### PR DESCRIPTION
**Pull request by emmettlu**
Fixs: #109

### CODE REVIEW

**Comment by aaravlu** [comment]
Could Claude please review this pr?

**Comment by deevus** [comment]
There's a bug that won't allow it to review PR's on another repo.

**Comment by deevus** [comment]
Please update the description with a detailed summary of what you changed and why. Thanks!

File: .github/workflows/test-integration.yml [hunk 1]
```diff
@@ -110,7 +110,7 @@ jobs:
       - name: Verify release files
         run: |
           # Check that all expected files exist
-          EXPECTED_FILES=("cli.exe" "cli.zip" "zed.exe" "zed.zip" "zed-opengl.exe" "zed-opengl.zip" "sha256sums.txt")
+          EXPECTED_FILES=("zed.zip" "zed-opengl.zip" "sha256sums.txt")
```
Comment (aaravlu) [inline .github/workflows/test-integration.yml]: We keep executables under `artificts` dir and make zip bundles under `release` dir.

File: .github/workflows/test-integration.yml [hunk 2]
```diff
@@ -135,22 +135,18 @@ jobs:
...
         scenario:
-          - name: "cli-only"
-            create_dx11: false
-            create_opengl: false
-            expected_files: 3
           - name: "dx11-only"
             create_dx11: true
             create_opengl: false
-            expected_files: 5
+            expected_files: 2
```
Comment (aaravlu) [inline .github/workflows/test-integration.yml]: count updating because we only count zip bundles

File: .github/workflows/test-integration.yml [hunk 3]
```diff
@@ -135,22 +135,18 @@ jobs:
...
           - name: "opengl-only"
             create_dx11: false
             create_opengl: true
-            expected_files: 5
+            expected_files: 2
```
Comment (aaravlu) [inline .github/workflows/test-integration.yml]: Same.

File: .github/workflows/test-integration.yml [hunk 4]
```diff
@@ -135,22 +135,18 @@ jobs:
...
           - name: "both-builds"
             create_dx11: true
             create_opengl: true
-            expected_files: 7
+            expected_files: 3
```
Comment (aaravlu) [inline .github/workflows/test-integration.yml]: Same.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -52,33 +52,11 @@ jobs:
           # Run the script
           ./scripts/prepare-release.sh
 
-          # Verify outputs
```
Comment (aaravlu) [inline .github/workflows/test.yml]: We only need `zed.zip`, `zed-opengl.zip`, and `sha256sums.txt` in `release` dir.

File: scripts/prepare-release.sh [hunk 6]
```diff
@@ -14,22 +14,29 @@ mkdir -p "$RELEASE_DIR"
 # Check if CLI build exists
 if [ -f "$ARTIFACTS_DIR/cli-release/cli.exe" ]; then
     echo "Found CLI build, adding to release..."
-    mv "$ARTIFACTS_DIR/cli-release/cli.exe" "$RELEASE_DIR/cli.exe"
-    zip -j "$RELEASE_DIR/cli.zip" -9 "$RELEASE_DIR/cli.exe"
+    mv "$ARTIFACTS_DIR/cli-release/cli.exe" "$ARTIFACTS_DIR/cli-release/zed.exe"
```
Comment (aaravlu) [inline scripts/prepare-release.sh]: rename `cli.exe` to `zed.exe`, no conflicts with GUI  editor's name because of operation in its `cli-release` dir

File: scripts/test-prepare-release.sh [hunk 7]
```diff
@@ -20,17 +20,17 @@ setup_test() {
     mkdir -p artifacts
 }
 
-create_dx11_artifact() {
+create_fake_dx11_artifact() {
```
Comment (aaravlu) [inline scripts/test-prepare-release.sh]: rename shell function name

File: scripts/test-prepare-release.sh [hunk 8]
```diff
@@ -144,19 +131,22 @@ else
 fi
 cd ..
 
-# Test 6: Verify zip files contain executables
```
Comment (aaravlu) [inline scripts/test-prepare-release.sh]: This test is no more needed

File: scripts/test-prepare-release.sh [hunk 9]
```diff
@@ -74,48 +74,35 @@ run_test() {
...
-create_opengl_artifact
+create_fake_cli_artifact
+create_fake_dx11_artifact
+create_fake_opengl_artifact
 run_test "success"
-verify_file_count 7  # cli.exe, cli.zip, zed.exe, zed.zip, zed-opengl.exe, zed-opengl.zip, sha256sums.txt
-verify_file_exists "release/cli.exe"
-verify_file_exists "release/cli.zip"
-verify_file_exists "release/zed.exe"
+verify_file_count 3
```
Comment (aaravlu) [inline scripts/test-prepare-release.sh]: `3`: `zed.zip`, `zed-opengl.zip`, `sha256sums.txt`

File: scripts/test-prepare-release.sh [hunk 10]
```diff
@@ -74,48 +74,35 @@ run_test() {
...
 
 # Test 2: Only DX11 build exists
 setup_test "Only DX11 build exists"
-create_dx11_artifact
+create_fake_cli_artifact
+create_fake_dx11_artifact
 run_test "success"
-verify_file_count 3  # zed.exe, zed.zip, sha256sums.txt
-verify_file_exists "release/zed.exe"
+verify_file_count 2
```
Comment (aaravlu) [inline scripts/test-prepare-release.sh]: `2`: `zed.zip`, `sha256sums.txt`

File: scripts/test-prepare-release.sh [hunk 11]
```diff
@@ -74,48 +74,35 @@ run_test() {
...
 
 # Test 3: Only OpenGL build exists
 setup_test "Only OpenGL build exists"
-create_opengl_artifact
+create_fake_cli_artifact
+create_fake_opengl_artifact
 run_test "success"
-verify_file_count 3  # zed-opengl.exe, zed-opengl.zip, sha256sums.txt
-verify_file_exists "release/zed-opengl.exe"
+verify_file_count 2
```
Comment (aaravlu) [inline scripts/test-prepare-release.sh]: `2`: `zed-opengl.zip`, `sha256sums.txt`

**Review by deevus — COMMENTED** [review]
# Code Review: PR #111 - Embed CLI into editor packages and update tests

## Summary

This PR attempts to fix issue #109 by embedding the CLI executable inside the editor zip packages with the correct directory structure. While the **intent and overall approach are correct**, there is a **critical bug** in the implementation that causes the zip files to contain incorrect paths.

## Critical Issue Found

### 🚨 CRITICAL: Incorrect Zip File Structure
**Location**: `scripts/prepare-release.sh:26, 35`

**Problem**: The zip command is being executed from the wrong directory, causing the zip file to include the full `artifacts/editor-*/` path prefix instead of just the `zed/` directory structure.

**Current behavior**:
```bash
zip -r "$RELEASE_DIR/zed.zip" -9 "$ARTIFACTS_DIR/editor-dx11-release/zed/"
```

This creates:
```
zed.zip
  `-- artifacts/
      `-- editor-dx11-release/
          `-- zed/
              `-- zed.exe
              `-- bin/
                  `-- zed.exe
```

**Required behavior** (from issue #109):
```
zed.zip
  `-- zed/
      `-- zed.exe (GUI editor)
      `-- bin/
          `-- zed.exe (CLI launcher)
```

**Recommended fix**:
```bash
# Check if DX11 build exists
if [ -f "$ARTIFACTS_DIR/editor-dx11-release/zed.exe" ]; then
    echo "Found DX11 build, adding to release..."

    mkdir -p "$ARTIFACTS_DIR/editor-dx11-release/zed/bin"
    mv "$ARTIFACTS_DIR/editor-dx11-release/zed.exe" "$ARTIFACTS_DIR/editor-dx11-release/zed"
    cp "$ARTIFACTS_DIR/cli-release/zed.exe" "$ARTIFACTS_DIR/editor-dx11-release/zed/bin"

    # Change to the directory containing 'zed/' before zipping
    (cd "$ARTIFACTS_DIR/editor-dx11-release" && zip -r "../../$RELEASE_DIR/zed.zip" -9 zed/)
fi
```

The same fix applies to the OpenGL build section (lines 30-36).

## Additional Issues

### 1. Test Coverage Gap
**Location**: `scripts/test-prepare-release.sh:139-158`

**Issue**: The zip content verification tests only check if `zed.exe` exists in the zip, but don't verify the **correct directory structure**. This is why the bug wasn't caught.

**Current test**:
```bash
if unzip -l release/zed.zip | grep -q "zed.exe"; then
    echo "✅ Zip file contains executable"
```

This passes even with the wrong directory structure (`artifacts/editor-dx11-release/zed/zed.exe`).

**Recommended improvement**:
```bash
# Test 6: Verify zip files contain correct directory structure
setup_test "Zip file content verification"
create_fake_cli_artifact
create_fake_dx11_artifact
create_fake_opengl_artifact
run_test "success"

# Check zed.zip structure - should have zed/ at root, not artifacts/...
if unzip -l release/zed.zip | grep -q "^.*zed/zed.exe$" && \
   ! unzip -l release/zed.zip | grep -q "artifacts/"; then
    echo "✅ zed.zip has correct structure"
else
    echo "❌ FAIL: zed.zip has incorrect directory structure"
    unzip -l release/zed.zip
    exit 1
fi

# Verify bin/zed.exe exists
if unzip -l release/zed.zip | grep -q "^.*zed/bin/zed.exe$"; then
    echo "✅ CLI executable found in correct location"
else
    echo "❌ FAIL: CLI executable missing or in wrong location"
    exit 1
fi
```

### 2. Potential Failure Mode Not Handled
**Location**: `scripts/prepare-release.sh:16-18, 24-26`

**Issue**: If the CLI build doesn't exist but editor builds do, the script will fail when trying to copy the non-existent CLI executable.

**Current code**:
```bash
# Check if CLI build exists
if [ -f "$ARTIFACTS_DIR/cli-release/cli.exe" ]; then
    echo "Found CLI build, adding to release..."
    mv "$ARTIFACTS_DIR/cli-release/cli.exe" "$ARTIFACTS_DIR/cli-release/zed.exe"
fi

# Check if DX11 build exists
if [ -f "$ARTIFACTS_DIR/editor-dx11-release/zed.exe" ]; then
    echo "Found DX11 build, adding to release..."
    # ... setup ...
    cp "$ARTIFACTS_DIR/cli-release/zed.exe" "$ARTIFACTS_DIR/editor-dx11-release/zed/bin"
    # ^ This will fail if CLI build doesn't exist!
```

**Recommended fix**: Make the CLI a required dependency for editor builds:
```bash
# Check if CLI build exists
CLI_EXISTS=false
if [ -f "$ARTIFACTS_DIR/cli-release/cli.exe" ]; then
    echo "Found CLI build, renaming..."
    mv "$ARTIFACTS_DIR/cli-release/cli.exe" "$ARTIFACTS_DIR/cli-release/zed.exe"
    CLI_EXISTS=true
fi

# Check if DX11 build exists
if [ -f "$ARTIFACTS_DIR/editor-dx11-release/zed.exe" ]; then
    if [ "$CLI_EXISTS" = false ]; then
        echo "Warning: DX11 build found but CLI build is missing. Skipping DX11 package."
    else
        echo "Found DX11 build, adding to release..."
        # ... rest of the code
    fi
fi
```

### 3. Test Scenario Missing
**Location**: `scripts/test-prepare-release.sh`

**Issue**: No test scenario for "editor build exists but CLI build doesn't exist". This edge case should be tested.

**Recommended addition**:
```bash
# Test: Editor build without CLI (should skip or handle gracefully)
setup_test "DX11 build exists but CLI build missing"
create_fake_dx11_artifact
# Don't create CLI artifact
run_test "success"  # or "failure" depending on desired behavior
# Verify appropriate handling
```

## Code Quality Assessment

### Positive Aspects
1. ✅ **Clean refactoring**: Removed standalone CLI zip generation correctly
2. ✅ **Good test naming**: `create_fake_*` functions are clearer than the original names
3. ✅ **Workflow updates**: GitHub Actions workflow changes are consistent with the new behavior
4. ✅ **Documentation**: Inline comments in PR explain the reasoning

### Areas for Improvement
1. ⚠️ **Directory navigation**: The script should use `cd` in subshells to avoid path prefixes in zip files
2. ⚠️ **Error handling**: No validation that CLI exists before editor builds try to use it
3. ⚠️ **Test assertions**: Tests should verify exact directory structure, not just file existence

## Workflow Changes Review

### `.github/workflows/test-integration.yml`
**Status**: ✅ Correct
- Removed expectations for standalone CLI and editor executables
- Updated file counts correctly (2 for single build + checksums, 3 for both builds + checksums)
- Removed "cli-only" scenario (correct decision since CLI is now bundled)

### `.github/workflows/test.yml`
**Status**: ✅ Correct
- Removed verification checks for individual executables
- Only verifies zip bundles exist

## Security Concerns

No security issues identified. The changes are limited to packaging logic.

## Performance Considerations

Using `zip -r` with `-9` (maximum compression) is appropriate for release artifacts. No performance concerns.

## Regression Risks

**HIGH RISK**: The critical bug means the zip files **do not meet the requirements** of issue #109. This will break:
- Any tools or documentation expecting the `zed/` directory at the zip root
- Manual extraction workflows that expect `zed/zed.exe` directly
- Scoop manifest updates mentioned in issue #109 will fail

## Recommendations

### Must Fix (Blocking)
1. ❌ **Fix zip directory structure** in `scripts/prepare-release.sh` lines 26 and 35-36
2. ❌ **Add structure validation** to tests in `scripts/test-prepare-release.sh`

### Should Fix (Important)
3. ⚠️ **Handle missing CLI gracefully** when editor builds exist
4. ⚠️ **Add test scenario** for editor-without-CLI case

### Nice to Have
5. 💡 Add a test that extracts the zip and verifies the exact file paths exist

## Overall Assessment

**REQUEST CHANGES** ⚠️

While the approach is sound and most of the implementation is correct, the critical bug in the zip creation logic means the PR does not achieve its stated goal of fixing issue #109. The zip files will have an incorrect directory structure that includes the `artifacts/editor-*/` path prefix.

The test coverage was improved but missed the structure validation that would have caught this bug.

### Action Items for Author

1. Update `scripts/prepare-release.sh`:
   - Line 26: Change zip command to use subshell with `cd`
   - Line 35-36: Same fix for OpenGL build

2. Update `scripts/test-prepare-release.sh`:
   - Add directory structure validation to zip content tests
   - Verify `zed/` is at the root (not `artifacts/...`)
   - Verify both `zed/zed.exe` and `zed/bin/zed.exe` exist

3. Test the fix:
   - Run the test suite
   - Manually extract the zip and verify the structure matches issue #109 requirements

---

**Files requiring changes:**
- `scripts/prepare-release.sh` (lines 22-36)
- `scripts/test-prepare-release.sh` (lines 138-158)

**Comment by aaravlu** [comment]
> I suggest using TDD and write a failing test that affirms the correct zip folder structure.

Sorry, i dont know how to do it.

**Comment by deevus** [comment]
Thanks mate. Just enjoying my weekend 

Will take a look soon 😘

**Comment by deevus** [comment]
@aaravlu Typo: lancher > launcher
